feat(ocr): tesseract support mis-oriented documents

ClemDoum opened the pull request:
<!-- Thank you for contributing to Docling! -->

<!-- STEPS TO FOLLOW:
  1. Add a description of the changes (frequently the same as the commit description)
  2. Enter the issue number next to "Resolves #" below (if there is no tracking issue resolved, **remove that section**)
  3. Make sure the PR title follows the **Commit Message Formatting**: https://www.conventionalcommits.org/en/v1.0.0/#summary.
  4. Follow the steps in the checklist below, starting with the **Commit Message Formatting**.
-->

<!-- Uncomment this section with the issue number if an issue is being resolved
**Issue resolved by this Pull Request:**
Resolves #
--->
# Description

This PR partially addresses #1155.

When input pages are rotated, Tesseract OCR fails to correctly parse them. However Tesseract supports OSD detection which allows to:
- get the document orientation
- rotate it if needed
- perform OCR with the document in its natural orientation
- then convert the bounding box back to the original document (rotated) orientation


## Known limitation

The PR only partially addresses the issues because the `LayoutModel` doesn't seem to support rotated page which causes some body texts to be detected as header and hence are not correctly output when converted to markdown or tokens (see debug outputs).

This seems unrelated to tesseract OCR output (which is correct when we use the debug mode to visulize the bounding boxes), hence I still think this contribution is valuable even if the output is not correct (see e2e tests). When the `LayoutModel` will correctly support rotated documents, we can expect a fully correct output.

## Potentially connected issues
- #683
- #639


**Checklist:**

- [x] Documentation has been updated, if necessary.
- [x] Examples have been added, if necessary.
- [x] Tests have been added, if necessary.

07pepa commented:
This also closes https://github.com/docling-project/docling/issues/639

cau-git commented on docling/models/tesseract_ocr_cli_model.py:
```diff
@@ -237,31 +249,30 @@ def __call__(
+                                rotated_image_size=high_res_image.size,
+                            )
 
                             cell = TextCell(
```
Since the `cell.rect` can now support arbitrary rectangle shapes (at an angle), it would be very useful to populate this rect with the correct orientation (instead of using `BoundingRectangle.from_bounding_box(...)`, which ignores any orientation.

cau-git reviewed (changes requested):
@ClemDoum This looks useful, even if isolated to the use of tesseract OCR. I left one remark below to make sure this change is reflected in the `TextCell` instances you create, because this information will be important downstream to understand if a cell is rotated or simply tall and narrow width.

Additionally, can you please rebase to current `main` and update all the test ground-truth? If the tests fail, you can run the line below, and observe which diff you get in the existing test data.

```
DOCLING_GEN_TEST_DATA=1 poetry run pytest ./tests/ -s -vv 
```

ClemDoum commented:
@cau-git  thank you for the review. I've:
- updated the tests
- updated the code to use proper `BoundingRectangle` keeping track of the angle (I was not 100% sure about the order  of `r_0`, `r_1`, `r_2`, `r_3`, I choose the so that if the box is a unit square, top left on the origin in top left coordinates, they are (0, 0), (1, 0), (1, 1), (0, 1))

 Additionally, do you want me to use the same rotation trick in the `LayoutModel` or are there some plans to make it rotation independent ? The trick could be something like:
 - detect of the rotation of the doc using majority voting (or average) of the `BoundingRectangle` angles or something like that)
 - rotate the doc
 - perform layout detection on the doc with the correct orientation
 
 The output of the `LayoutPredictor` doesn't seem to contain much rotation dependent stuff so I guess I wouldn't have to revert anything from the output.

ClemDoum commented:
And sorry for amending the commit, I should have created a new one, it would have been easier to see the post review diff

cau-git commented:
@ClemDoum Many thanks for the updates. I missed to point you to the BoundingRectangle convention, we define it like this for an upright rectangle.

```
      rx_3, ry_3        rx_2, ry_2
         +----------------+
         |                             |
         |                             |
         |                             |
         +----------------+
      rx_0, ry_0        rx_1, ry_1
```
That is, the (rx_0, ry_0) is the bottom left corner, no matter in which `CoordOrigin`, and then the coordinates go counter-clockwise.

cau-git commented:
> Additionally, do you want me to use the same rotation trick in the `LayoutModel` or are there some plans to make it rotation independent ?

Yes, we would like to adopt the rotation correction you implement for tesseract here into the Layout Model and the TableFormer model, in the way you suggested (through a heuristic that determines the majority direction of the text cells).
Would you be willing to help on this? @maxmnemonic will give you pointers in case.

maxmnemonic commented:
@ClemDoum, as @cau-git mentioned, if you want to help us with cluster rotations, you are very welcome!
We have a way to determine cluster orientation from orientation of each underlying PDF cells with new (default) PDF backend.
And here is an example on how you can do it: https://github.com/docling-project/docling/pull/1286/files

Please let us know if you would like to proceed with this, thanks!

ClemDoum commented:
@maxmnemonic I've updated the `LayoutModel`, however unit test are still not looking good because I think there's some bugs in the `BoundingRectangle.angle_360` (and maybe also `angle`).  When we get an `tan(thêta) = 0` which is `delta_y = 0` we consider online one of the possible solutions (`0`) however the solution are only unique between `[-pi / 2, pi /2]`. Since we consider solution between `[0, 2pi]` (or [-pi, pi]), there are other solutions and we need to consider the `delta_x` to find the right one.

I think it should be something like this instead:
```python
@property
    def angle_360(self) -> int:
        """Calculate the angle of the rectangle in degrees (0-360 range)."""
        p_0 = ((self.r_x0 + self.r_x3) / 2.0, (self.r_y0 + self.r_y3) / 2.0)
        p_1 = ((self.r_x1 + self.r_x2) / 2.0, (self.r_y1 + self.r_y2) / 2.0)

        delta_x, delta_y = p_1[0] - p_0[0], p_1[1] - p_0[1]

        if abs(delta_y) < 1.0e-2:
            angle =  0
        elif abs(delta_x) < 1.0e-2:
            angle = 90
        else:
            angle = round(-math.atan(delta_y / delta_x) / np.pi * 180)
        if delta_x < 0:
            angle += 180
        return angle
```

ClemDoum commented:
@maxmnemonic concerning the `TableStructureModel` could you confirm that the steps to support rotation would be:
1. detect if the page is rotated using it cells
2. rotate it to the natural orientation
3. rotate the `tbl_box`

But then I'm a bit lost:
- I guess I should rotate the `page_input["tokens"]["bbox"]` accordlingly ?
- should i also reorder the `page_input["tokens"]` themselves (are they expected to be in some particular order ?) ?
- then I guess once the `tf_predictor` has made prediction, I guess I should rotate the results back to the original image orientation ?

cau-git commented:
> @cau-git I've updated the `LayoutModel`, however unit test are still not looking good because I think there's some bugs in the `BoundingRectangle.angle_360` (and maybe also `angle`). When we get an `tan(thêta) = 0` which is `delta_y = 0` we consider online one of the possible solutions (`0`) however the solution are only unique between `[-pi / 2, pi /2]`. Since we consider solution between `[0, 2pi]` (or `[-pi, pi]`), there are other solutions and we need to consider the `delta_x` to find the right one.
> 
> I think it should be something like this instead:
> ...


@ClemDoum Many thanks. Can you make a companion PR on `docling-core` where you update [the angle_360 implementation](https://github.com/docling-project/docling-core/blob/main/docling_core/types/doc/page.py#L132-L145) as you suggest?  Many thanks.

maxmnemonic commented:
> @maxmnemonic concerning the `TableStructureModel` could you confirm that the steps to support rotation would be:
> 
> 1. detect if the page is rotated using it cells
> 2. rotate it to the natural orientation
> 3. rotate the `tbl_box`
> 
> But then I'm a bit lost:
> 
> * I guess I should rotate the `page_input["tokens"]["bbox"]` accordlingly ?
> * should i also reorder the `page_input["tokens"]` themselves (are they expected to be in some particular order ?) ?
> * then I guess once the `tf_predictor` has made prediction, I guess I should rotate the results back to the original image orientation ?

It could be that the page is oriented "naturally" but table on the page rotated sideways, or could be another combination.
But for every rotation of the image, we should rotate bounding boxes too. Please keep the order, as it indexes and binds all of the elements together, so:

1. rotate page - Yes
2. reorder - No
3. rotate input for TableFormer, when needed (input text cells and table image) - Yes; here I assume after prediction and matching with underlying text cells are done, then we don't need to do any extra transformations as long as order of the cells remain the same.

ClemDoum commented:
@cau-git I've made a PR on docling-core: https://github.com/docling-project/docling-core/pull/237

ClemDoum commented:
Thanks for the explanation @maxmnemonic I'll try to wrap it up tomorrow !

cau-git commented:
@ClemDoum To be sure we don't need to delay merges, let's separate out concerns that do not relate to the tesseract orientation handling.

I would suggest you create another PR for:
- detect_orientation utility
- fixes in the layout model debug viz
- anything beyond

Then, we can review and test the purely tesseract-related work in isolation.

ClemDoum commented:
@cau-git I've removed the table debug stuff and created another branch for anything unrelated to tesseract.

I've kept the orientation utility separate from the `ocr_utils.py` and refactored it to be usable for other purposes (I 
needed it as well to rotate bounding boxes for the table detection and the previous API wasn't convenient at all)

ClemDoum commented:
@cau-git are you OK with the latest updates ? Should I update something else ?

ClemDoum commented:
@cau-git done

cau-git commented:
@ClemDoum Thanks for the update. Since the test data changed again through a PR merge to `main` just earlier today, the test case re-generation has to run once more.
```
DOCLING_GEN_TEST_DATA=1 poetry run pytest ./test -s -vv 
```

I was trying to see if I can cover it, but I can not push back to your personal fork. Either you run it once more, or I transfer this fork into `docling-project/docling` to fix it there (and then open a new PR for it). Let me know what you prefer. Thanks

cau-git reviewed (approved):
LGTM

maxmnemonic reviewed (approved):
Big piece of work! Thanks @ClemDoum!

nikos-livathinos commented on docling/models/tesseract_ocr_cli_model.py:
```diff
@@ -225,8 +231,30 @@ def __call__(
+                                    image_file,
+                                    exc.stderr,
+                                )
+                                continue
```
Seeing that pages are skipped when OSD fails, I wonder if OCR can still produce correct results despite the failure of OSD. Do we have any insight on that?

For example, another approach to use more document pages, would be:
1. Assuming the user has NOT provided the `auto` language.
2. If OSD fails, we skip the auto-rotation step and we proceed with OCR.

ClemDoum replied:
You are right, we do OCR using the default Page Segmentation Mode (PSM) which is `--psm 3`.

According to the [documentation](https://tesseract-ocr.github.io/tessdoc/Command-Line-Usage.html) the Orientation and Script Detection (OSD) is not run again so in theory it's possible that OCR succeeds. 

If the Script detection fails I don't really see how the OCR could output anything without first picking up a script so I doubt OCR will succeed, but if the orientation detection fails, it might still be possible to output something relevant.

Updating the PR with your strategy

ClemDoum replied:
I've updated the PR, I also made sure to use `subprocess.run` instead of `Popen.communicate` otherwise we don't get any `CalledProcessError` (errors are just capture in `stdout`/`stderr`, which will make code following the python code fail in most cases)

nikos-livathinos commented on docling/models/tesseract_ocr_cli_model.py:
```diff
@@ -115,13 +119,12 @@ def _run_tesseract(self, ifilename: str):
 
-        proc = Popen(cmd, stdout=PIPE, stderr=DEVNULL)
-        output, _ = proc.communicate()
+        output = subprocess.run(cmd, stdout=PIPE, stderr=DEVNULL)
```
I think we should set the parameter `check=True`

ClemDoum replied:
You're right thanks, fixing it !

nikos-livathinos commented on docling/models/tesseract_ocr_cli_model.py:
```diff
@@ -182,9 +187,8 @@ def _set_languages_and_prefix(self):
-        proc = Popen(cmd, stdout=PIPE, stderr=DEVNULL)
-        output, _ = proc.communicate()
-        decoded_data = output.decode("utf-8")
+        output = subprocess.run(cmd, stdout=PIPE, stderr=DEVNULL)
```
I think we should set the parameter `check=True`

nikos-livathinos commented on docling/models/tesseract_ocr_model.py:
```diff
@@ -139,16 +144,17 @@ def __call__(
+                        osd = self.osd_reader.DetectOrientationScript()
+                        # No text, probably
+                        if osd is None:
+                            continue
```
Here we skip the entire page when OSD fails.
I think we should change the logic as it is in `TesseractOcrCliModel`. See
https://github.com/ClemDoum/docling/blob/9d8e15a8545bd9af94013b5d0a6771dd00f25492/docling/models/tesseract_ocr_cli_model.py#L239

Overall, when OSD fails:
1. If `auto` is set in the languages, we should skip the page (because we cannot detect the language without OSD).
2. If `auto` is not set, we should log an error message and proceed with the OCR.

cau-git commented on docling/models/tesseract_ocr_cli_model.py:
```diff
@@ -54,6 +54,7 @@ def __init__(
         self._version: Optional[str] = None
         self._tesseract_languages: Optional[List[str]] = None
         self._script_prefix: Optional[str] = None
+        self._is_auto: bool
```
for safety, can you please assign a default value `False` (then the later assignment in `def _set_languages_and_prefix()`can overwrite it)

cau-git reviewed (commented):
@ClemDoum thanks for adding the `_is_auto` checks, the only comment I have is below, otherwise all looks good to me.

cau-git commented:
@ClemDoum The CI tests reveal some failure now on tesseract-related test units, could you have a check there too? Many thanks

ClemDoum commented:
Thank you, well caught @cau-git, I used the wrong property to compute the flag value, tests are now green again locally

nikos-livathinos commented on docling/models/tesseract_ocr_model.py:
```diff
@@ -139,16 +143,28 @@ def __call__(
+                                conv_res.input.file,
+                                page_i,
+                                ocr_rect_i,
+                            )
```
Maybe better to move this log message before the `if self._is_auto:  continue`, so that the user can always see it, regardless if we skip the page or not.

ClemDoum replied:
You're right, updated

nikos-livathinos reviewed (approved):
Very good job!

cau-git reviewed (approved):
🚀 Great contribution @ClemDoum! 
Time to 🚢
